### PR TITLE
implement external functions

### PR DESCRIPTION
**Pull request by CrowdHailer**
tackles #1049

### CODE REVIEW

File: src/javascript.rs [hunk 1]
```diff
@@ -111,6 +118,40 @@ fn module_function<'a>(
...
+    arguments: &'a [ExternalFnArg<T>],
+    module: &'a str,
+    fun: &'a str,
+) -> Output<'a> {
+    let head = if public {
+        "export function "
+    } else {
+        "function "
+    };
+    let arguments = wrap_args(arguments.iter().enumerate().map(|a| match a {
```
Comment (CrowdHailer) [inline src/javascript.rs:133]: I tried to look for a clever way to reuse some of the code in `module_function` but in the end just keeping it separate seemed clearest.

File: src/javascript/tests.rs [hunk 2]
```diff
@@ -174,3 +174,30 @@ const e = [
...
+
+function foo() {
+  document.foo()
+}
+
+export function bar(arg0, arg1) {
+  document.bar(arg0, arg1)
+}
+
+function baz(x, y) {
```
Comment (CrowdHailer) [inline src/javascript/tests.rs:198]: I use the name of the arguments if they are given, to make the output nicer.

Comment (lpil) [reply src/javascript/tests.rs:198]: Very nice

File: src/javascript.rs [hunk 3]
```diff
@@ -111,6 +118,40 @@ fn module_function<'a>(
...
+        (index, ExternalFnArg { label, .. }) => {
+            let arg_name = label.clone().unwrap_or(format!("arg{}", index));
+            Document::String(arg_name)
+        }
```
Comment (lpil) [inline src/javascript.rs:137]: I think we can remove the cloning of the name here

```suggestion
        (index, ExternalFnArg { label, .. }) => {
            label.as_ref().map(Documentable::to_doc).unwrap_or_else(|| Document::String(format!("arg{}", index)));
        }
```

File: src/javascript.rs [hunk 4]
```diff
@@ -111,6 +118,40 @@ fn module_function<'a>(
...
+    }));
+    let body = module
+        .to_doc()
+        .append(".")
+        .append(fun)
+        .append(arguments.clone());
```
Comment (lpil) [inline src/javascript.rs:143]: Need a `return` too. Pesky statement based language.
```suggestion
    let body = docvec!["return ", module, ".", fun, arguments.clone()];
```

File: src/javascript.rs (hunk 1 above)
Comment (lpil) [reply src/javascript.rs:133]: I like this aye

**Review by lpil — COMMENTED** [review]
Looks good! Could you rebase on main? I changed the structure a tiny bit to hold some extra state

File: src/javascript.rs (hunk 3 above)
Comment (CrowdHailer) [reply src/javascript.rs]: Have added and got.

```
the trait `pretty::Documentable<'_>` is not implemented for `&std::string::String`
```

File: src/javascript.rs [hunk 5]
```diff
@@ -119,6 +126,41 @@ impl<'a> Generator<'a> {
...
+        let head = if public {
+            "export function "
+        } else {
+            "function "
+        };
+        let arguments = wrap_args(arguments.iter().enumerate().map(|a| match a {
+            (index, ExternalFnArg { label, .. }) => {
+                let arg_name = label.clone().unwrap_or(format!("arg{}", index));
+                Document::String(arg_name)
+                // label
```
Comment (CrowdHailer) [inline src/javascript.rs:147]: This is the code, from you suggestion, moved after merging main. I wasn't sure how to use it so have left my original code as it was functioning.

File: src/javascript/tests.rs [hunk 6]
```diff
@@ -181,3 +181,30 @@ const e = [
...
+pub external type Thing
+external fn foo() -> Thing = "document" "foo"
+pub external fn bar(Int, String) -> Thing = "document" "bar"
+external fn baz(x: Int, y: String) -> Thing = "document" "baz"
+ 
+"#,
+        r#""use strict";
+
+function foo() {
+  return document.foo()
```
Comment (lpil) [inline src/javascript/tests.rs:198]: Oh, I just had a thought. We're probably going to have to `import { foo } from "document";` ?

Comment (CrowdHailer) [reply src/javascript/tests.rs:198]: No, don't think so. My example project I had working in the browser didn't require any of that.
I guess bundlers might need separate thinking about, but in those cases it's probably a setting in the bundler to flag what globals are referenced.

Comment (lpil) [reply src/javascript/tests.rs:198]: Let's do this in a later PR

**Review by lpil — APPROVED** [review]

File: src/javascript.rs (hunk 3 above)
Comment (lpil) [reply src/javascript.rs]: OK, will need to turn the string into a str first
```rust
label.as_ref().map(|l| l.as_str().to_doc()).unwrap_or_else(|| Document::String(format!("arg{}", index)));
```

File: src/javascript.rs (hunk 5 above)
Comment (lpil) [reply src/javascript.rs:147]: This would be the new definition of `arg_name`

**Comment by lpil** [comment]
Could you resolve the merge conflicts? Seems to be disabled for me

File: src/javascript/tests.rs (hunk 6 above)
Comment (lpil) [reply src/javascript/tests.rs:198]: Actually I just had an idea how how externals could work that may work well and means we don't need to do this

**Comment by CrowdHailer** [comment]
There seems to be a problem with the option to allow you to edit as I am using the midas-framework org. https://github.com/isaacs/github/issues/1681 
Will do it myself now

**Pull request merged**